### ISSUE TITLE
response.status sometimes 0 when using ReadableStream 

### ISSUE THREAD

**Issue by fabiancook** [oven-sh/bun#1190]
### Version

0.1.11

### Platform

Linux fv-az246-574 5.15.0-1017-azure #20~20.04.1-Ubuntu SMP Fri Aug 5 12:16:53 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### What steps will reproduce the bug?

Using async iterator + ReadableStream + Response + fetch

Bun version:

```javascript
let timeout = 0;
async function* generate() {
    yield "[";
    yield JSON.stringify({ a: "b" });
    for (let i = 0; i < 3; i += 1) {
        yield ",";
        yield JSON.stringify({ i });
    }
    yield "]";
}
function toPullUnderlyingSource(iterable) {
    const encoder = new TextEncoder();
    let iterator;
    return {
        start() {
            iterator = iterable[Symbol.asyncIterator]();
        },
        async pull(controller) {
            const { value, done } = await iterator.next();
            if (done) {
                if (timeout) {
                    await new Promise(resolve => setTimeout(resolve, timeout));
                }
                controller.close();
            }
            else {
                controller.enqueue(encoder.encode(value));
            }
        },
        async cancel() {
            await iterator.return();
        }
    };
}
export function toStream(iterable) {
    const source = toPullUnderlyingSource(iterable);
    return new ReadableStream(source);
}
export function toResponse(iterable) {
    return new Response(toStream(iterable), {
        headers: {
            "Content-Type": "application/json",
            "X-Content-Type-Options": "nosniff"
        }
    });
}
const server = Bun.serve({
    fetch() {
        return toResponse(generate());
    },
    port: 3010
});
const url = "http://0.0.0.0:3010"
console.log(`HTTP webserver running. Access it at: ${url}`);
async function run(loops = 10) {
    let ok = 0, notOk = 0, exitCode = 0;
    for (let i = 0; i < loops; i += 1) {
        const response = await fetch(url);
        if (!response.ok) {
            console.error("Response not ok");
            console.error(response);
            notOk += 1;
            exitCode = 1;
        }
        else {
            console.log("Response ok");
            console.log(response);
            ok += 1;
        }
    }
    return {
        ok,
        notOk,
        exitCode
    };
}
const timeouts = [
    0,
    // 100,
    // 500,
    // 1500,
    // 3000,
    // 10000
];
let allNotOk = 0, allOk = 0, exitCode = 0, maxTimeout = 0;
for (const currentTimeout of timeouts) {
    timeout = currentTimeout;
    maxTimeout = currentTimeout;
    const result = await run(500);
    console.log(result);
    allNotOk += result.notOk;
    allOk += result.ok;
    exitCode = exitCode || result.exitCode;
    if (!result.notOk) {
        // We didn't run into the problem on this platform with this timeout
        break;
    }
}
console.log({ allNotOk, allOk, maxTimeout });
await server.stop();
if (exitCode) {
    process.exit(exitCode);
}
```

Deno version:

```javascript
const server = Deno.listen({ port: 0 });
const hostname = `http://0.0.0.0:${server.addr.port}`;
console.log(`HTTP webserver running. Access it at: ${hostname}`);
const abortController = new AbortController();
let timeout = 0;
async function* generate() {
    yield "[";
    yield JSON.stringify({ a: "b" });
    for (let i = 0; i < 3; i += 1) {
        yield ",";
        yield JSON.stringify({ i });
    }
    yield "]";
}
function toPullUnderlyingSource(iterable) {
    const encoder = new TextEncoder();
    let iterator;
    return {
        start() {
            iterator = iterable[Symbol.asyncIterator]();
        },
        async pull(controller) {
            const { value, done } = await iterator.next();
            if (done) {
                if (timeout) {
                    await new Promise(resolve => setTimeout(resolve, timeout));
                }
                controller.close();
            }
            else {
                controller.enqueue(encoder.encode(value));
            }
        },
        async cancel() {
            await iterator.return();
        }
    };
}
export function toStream(iterable) {
    const source = toPullUnderlyingSource(iterable);
    return new ReadableStream(source);
}
export function toResponse(iterable) {
    return new Response(toStream(iterable), {
        headers: {
            "Content-Type": "application/json",
            "X-Content-Type-Options": "nosniff"
        }
    });
}
const onComplete = (async function watch() {
    for await (const connection of server) {
        void handleConnection(connection).catch(error => void error);
    }
    async function handleConnection(connection) {
        const http = Deno.serveHttp(connection);
        abortController.signal.addEventListener("abort", () => {
            try {
                http.close();
            }
            catch { }
        });
        for await (const event of http) {
            event.respondWith(toResponse(generate()))?.catch?.(error => void error);
        }
    }
})();
async function run(loops = 10) {
    let ok = 0, notOk = 0, exitCode = 0;
    for (let i = 0; i < loops; i += 1) {
        const response = await fetch(hostname);
        if (!response.ok) {
            console.error("Response not ok");
            console.error(response);
            notOk += 1;
            exitCode = 1;
        }
        else {
            console.log("Response ok");
            console.log(response);
            ok += 1;
        }
    }
    return {
        ok,
        notOk,
        exitCode
    };
}
const timeouts = [
    0,
    // 100,
    // 500,
    // 1500,
    // 3000,
    // 10000
];
let allNotOk = 0, allOk = 0, exitCode = 0, maxTimeout = 0;
for (const currentTimeout of timeouts) {
    timeout = currentTimeout;
    maxTimeout = currentTimeout;
    const result = await run(500);
    console.log(result);
    allNotOk += result.notOk;
    allOk += result.ok;
    exitCode = exitCode || result.exitCode;
    if (!result.notOk) {
        // We didn't run into the problem on this platform with this timeout
        break;
    }
}
console.log({ allNotOk, allOk, maxTimeout });
server.close();
abortController.abort();
await onComplete.catch(error => void error);
if (exitCode) {
    process.exit(exitCode);
}
```

[GitHub Repo/Commit](https://github.com/virtualstate/focus/tree/15166772e410053eea235aab7c19d172dd055d7a)
[GitHub File](https://github.com/virtualstate/focus/blob/15166772e410053eea235aab7c19d172dd055d7a/src/tests/server/bun-issue-reproduction.ts)

Bun GitHub action: 

```yml
name: bun-test-server-actions
on: [push, pull_request]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - uses: actions/setup-node@v2
        with:
          node-version: "18.x"
          registry-url: "https://registry.npmjs.org"
      - uses: antongolub/action-setup-bun@v1
      - run: |
          yarn install
          bun install
      - run: yarn build
      - run: bun esnext/tests/server/bun-issue-reproduction.js
```

Deno GitHub action: 

```yml
name: deno-test-server-actions
on: [push, pull_request]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - uses: actions/setup-node@v2
        with:
          node-version: "18.x"
          registry-url: "https://registry.npmjs.org"
      - uses: denoland/setup-deno@v1
        with:
          deno-version: "v1.x"
      - run: yarn install
      - run: yarn build
      - run: deno run --allow-net esnext/tests/server/deno-issue-reproduction.js
```

### How often does it reproduce? Is there a required condition?

In each process I ran `fetch` 500 times, below are the counts where `response.ok` was `false` 

| Run | [Bun GitHub Actions](https://github.com/virtualstate/focus/runs/8133746144?check_suite_focus=true#step:7:5010) | Bun Mac M1 | [Deno GitHub Actions](https://github.com/virtualstate/focus/runs/8133749967?check_suite_focus=true#step:7:8509) | Deno Mac M1
| ------|---------------------|----------------|-----------|------------|
| 1 | 18 / 500 | 45 / 500  | 0 / 500 | 0 / 500 | 
| 2 | 11 / 500 | 48 / 500 | 0 / 500 | 0 / 500 | 
| 3 | 13 / 500 | 57 / 500  | 0 / 500 | 0 / 500 | 
| 4 | 17 / 500 | 78 / 500 | 0 / 500 | 0 / 500 | 
| 5 | 10 / 500 | 42 / 500 | 0 / 500 | 0 / 500 | 
| Avg | 2.7% | 10.8% | | 

### What is the expected behavior?

[Produce response.ok](https://github.com/virtualstate/focus/runs/8133746144?check_suite_focus=true#step:7:540)

```
Response ok
Response (35 bytes) {
  ok: true,
  url: "http://localhost:3010/",
  statusText: "HM",
  redirected: false,
  bodyUsed: false,
  status: 200,
  Blob (35 bytes)
}
```

### What do you see instead?

[Produces response.ok = false](https://github.com/virtualstate/focus/runs/8133746144?check_suite_focus=true#step:7:550)

```
Response not ok
Response (0 KB) {
  ok: false,
  url: "http://localhost:3010/",
  statusText: "",
  redirected: false,
  bodyUsed: false,
  status: 0,
  [Blob detached]
}
```

### Additional information

I came across this issue while [running tests in GitHub actions](https://github.com/virtualstate/focus/runs/8133638295?check_suite_focus=true#step:11:1211) where this combination of code was used. I would run into the issue sometimes 3/5 times, locally on an m1 though I hadn't ran into this issue until creating the above script to try over and over. 

I also tried a version where an iterator is used (rather than asyncIterator), and where pull doesn't return a promise, but I still run into `!response.ok`

**Comment by Jarred-Sumner** [comment]
Fixed in Bun v0.2.0 or Bun v0.1.13

**Comment by fabiancook** [comment]
```shell
Response ok
Response (35 bytes) {
  ok: true,
  url: "http://0.0.0.0:3010",
  statusText: "HM",
  redirected: false,
  bodyUsed: false,
  status: 200,
  Blob (35 bytes)
}
{ ok: 500, notOk: 0, exitCode: 0 }
{ allNotOk: 0, allOk: 500, maxTimeout: 0 
```

👍🏻  Confirmed fixed on my end.
